File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        cntr = {}
        # 1-index
        bucket = {i: set() for i in range(n + 2)}
        max_freq = 0

        for num in nums:
            if num not in cntr:
                cntr[num] = 1
                bucket[1].add(num)
            bucket[cntr[num]].remove(num)
            cntr[num] += 1
            bucket[cntr[num]].add(num)
            max_freq = max(
                max_freq,
                cntr[num]
            )
        
        res = []
        while k and max_freq:
            if bucket[max_freq]:
                res.extend(bucket[max_freq])
                k -= len(bucket[max_freq])
            max_freq -= 1
        
        return res

File: Data Structures & Algorithms/contains-duplicate-ii/submission-2.cs
public class Solution {
    public bool ContainsNearbyDuplicate(int[] nums, int k) {
        var size = nums.Length;
        if (size == 1 || k == 0)
        {
            return false;
        }

        var vis = new HashSet<int>();

        var l = 0;
        for (int r = 0; r < size; r++)
        {
            if (vis.Contains(nums[r]))
            {
                return true;
            }
            vis.Add(nums[r]);

            if (r - l == k)
            {
                vis.Remove(nums[l++]);
            }
        }

        return false;

    }
}